lib/components/element.rb

```diff
@@ -45,20 +51,28 @@ def self.element(name, multiple: false, &config)
         get_instance_variable(plural_name)
       end
     end
+
     # rubocop:enable Metrics/AbcSize
     # rubocop:enable Metrics/CyclomaticComplexity
     # rubocop:enable Metrics/MethodLength
     # rubocop:enable Metrics/PerceivedComplexity
 
+    attr_reader :content_block
```

Comment:
Should we document this as well as `content_block?` somewhere? I'm not sure if it belongs under the validation section or perhaps earlier, under "Attributes and blocks". I suppose they could be used for other things than validation. What do you think?

Do you think `block_content` is clearer than `content_block` btw? As in, "the content of the block", since the variable doesn't actually contain a block, but rather the yielded content of the block.

Reply:
👍 on `block_content` - I already renamed it.

I was wondering if we should use the 'block_content' reader as preferred way of rendering blocks in view templates - instead of rendering the component instance via the to_s method. It clarifies its output and aligns with other methods and attribute calls we make on the instance (e.g. alert.context or alert.block_content?). In this case we should probably deprecate the to_s method?

```ruby
<% # app/components/alert/_alert.html.erb %>

<div class="alert alert--<%= alert.context %>" role="alert">
  <%= alert.block_content %>
</div>
```

instead of

```ruby
<% # app/components/alert/_alert.html.erb %>

<div class="alert alert--<%= alert.context %>" role="alert">
  <%= alert %>
</div>
```

Reply:
I have a strong interest in moving away from `to_s` as well, although I would love to see elements either dump their content or support a render method so that they can specify their own templating if need be.

Reply:
It's a valid question for sure. The reason I did it in the first place was because I couldn't come up with a good name, that wouldn't potentially clash with an attribute, and that felt short enough to type over and over in a lot of components.

Perhaps let's keep `to_s` for now, to remain backwards compatible. At least there is now an alternative with `block_content`.

@imathis Probably a separate discussion, but what would be the reason for specifying templates for elements? You could always use a separate component, but perhaps you still want them to be internal to the component, and not usable from outside? I suppose you could always implement this by defining a helper:

```ruby
class Component
  element :body

  def render_body
    @view.render(partial: "../path/to/body/partial", object: body)
  end
end

Reply:
@jensljungblad should I update the README to use `alert.content_block` als default then and mention the "instance to_s" rendering as "shortcut"? We could also not mention the "instance to_s" rendering, still support it and add a deprecation warning via log file? Describing two way to render the components block seems a bit awkward to me ...

Reply:
I guess I don't really care about backwards compatibility at this point. It's still not even a released gem :) Sure, let's remove `#to_s` and update the README to use `#block_content` instead. More verbose for sure but perhaps makes more sense. And can always be revisited down the line if someone has a better idea.

Reply:
After eliminating the to_s method and using the block_content method for block rendering instead, I get your point on verbosity. Especially the rendering of "simple" elements (without further attributes) appears  a bit bulky (see changes on component views in dummy/app)

Maybe a shorter name is appropriate. We could just call it `block` instead of `block_content` (and `block?` instead of `block_content?`)?

Reply:
Or maybe `content` isn't a terrible name. It's close to Rails own `content_for` helpers, which do a similar thing: https://api.rubyonrails.org/classes/ActionView/Helpers/CaptureHelper.html#method-i-capture. I suppose I ended up using `to_s` so I wouldn't have to name this thing :)

I suppose it could be a problem if you have an attribute named `content`. But maybe that's not a big deal.

Another alternative is to merge validations, and create a new issue for adding a reader/moving away from `to_s`. You could implement the `block_content?` method yourself in the meanwhile:

```ruby
class Component
  def block_content?
    @yield.present?
  end
end
```

Reply:
I will say I nearly used `content` as an attribute for an attribute once, and I can imagine it's quite a reasonable choice. Having to share space with an end-user's attributes and elements puts any choice in danger of breaking things. I've been working on a PR which adds features to elements, but fear of breaking things due to naming conflicts makes it hard to expect upstream adoption. Is there a way to protect collisions for internal instance methods or attributes? Like (forgive me) underscoring internal methods, e.g., `_content`.

Reply:
I just reviewed our project again and noticed that we have been building several components with an `content` attribute by instinct 🤓

Reply:
Both names have their pros and cons. From my perspective some content gets injected into the component by a ruby block, therefore it somehow feels natural to reference that block by calling `block`. 

@jensljungblad I guess the final naming is your call ;) Feel free to rename as you like and lets 🚀

Reply:
Let's extract this discussion from this PR, it feels like a bit of an aside, and not crucial to validations. I created an issue here https://github.com/jensljungblad/components/issues/32 where we can discuss more about the block and its API.

If you want we can just remove the block stuff from this PR and merge it, and we'll have validations for attributes ready for now.

Reply:
All changes concerning the block handling have been reverted.